src/ray/gcs/gcs_client/service_based_accessor.cc

```diff
@@ -549,22 +597,67 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
     const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Subscribing batch heartbeat.";
   RAY_CHECK(subscribe != nullptr);
-  auto on_subscribe = [subscribe](const ClientID &node_id,
-                                  const HeartbeatBatchTableData &data) {
-    subscribe(data);
+  auto on_subscribe = [subscribe](const std::string &id, const std::string &data) {
+    HeartbeatBatchTableData heartbeat_batch;
+    heartbeat_batch.ParseFromString(data);
+    subscribe(heartbeat_batch);
   };
-  auto status = heartbeat_batch_sub_executor_.AsyncSubscribeAll(ClientID::Nil(),
-                                                                on_subscribe, done);
+  auto status = gcs_sub_.SubscribeAll(heartbeat_batch_channel_, on_subscribe, done);
   RAY_LOG(DEBUG) << "Finished subscribing batch heartbeat.";
   return status;
 }
 
+void ServiceBasedNodeInfoAccessor::HandleNotification(const GcsNodeInfo &node_info) {
+  ClientID node_id = ClientID::FromBinary(node_info.node_id());
+  bool is_alive = (node_info.state() == GcsNodeInfo::ALIVE);
+  // It's possible to get duplicate notifications from the client table, so
```

Comment:
why would we receive duplicate notifications?

Reply:
fixed